frontend/src/helpers/dataOverlay.js

```diff
@@ -0,0 +1,97 @@
+const DEFAULT_GENE_COLOR = '#feb';
+const DEFAULT_METABOLITE_COLOR = '#9df';
+
+const DATA_TYPES_COMPONENTS = {
+  transcriptomics: {
+    componentType: 'gene',
+    className: 'enz',
+    defaultColor: DEFAULT_GENE_COLOR,
+  },
+  metabolomics: {
+    componentType: 'metabolite',
+    className: 'met',
+    defaultColor: DEFAULT_METABOLITE_COLOR,
+  },
+};
+
+/*
+ * @param { Blob } file
+ *
+ * @example
+ * try {
+ *   const data = await parseCustomFile(file);
+ * } catch (e) {
+ *   console.error(e)
+ * }
+ */
+const parseFile = async (file) => {
+  let tissues = [];
+  const levels = { 'n/a': 'n/a' };
+
+  const text = await file.text();
+
+  const lines = text.split(/\r?\n/);
+  let indexLine = 1;
+  // fetch Tissue
+  if (lines[0].split('\t').length !== 1) {
+    const arrLine = lines[0].split('\t');
+    const v = Number(arrLine[1]);
+    if (Number.isNaN(v)) {
+      tissues = lines[0].split('\t');
+      tissues.shift();
+      lines.shift();
+    } else {
+      tissues = [];
+      for (let i = 1; i < arrLine.length; i += 1) {
+        tissues.push(`serie${i}`);
+      }
+    }
+  } else {
+    throw new Error('Invalid TSV format, expect at least two columns.');
+  }
+
+  // parse lines
+  // make tissues key;
+  for (let i = 0; i < tissues.length; i += 1) {
+    const tissue = tissues[i];
+    if (tissue in levels) {
+      throw new Error(`Error: duplicated column '${tissue}'.`);
+    }
+    levels[tissue] = {};
+  }
+
+  let entriesCount = 0;
+  for (let k = 0; k < lines.length; k += 1) {
+    const line = lines[k];
+    if (line) {
+      const arrLine = line.split('\t');
+      if (arrLine.length !== tissues.length + 1) {
+        throw new Error(`Error: invalid number of values line ${indexLine}.`);
+      }
+      for (let i = 1; i < arrLine.length; i += 1) {
+        if (arrLine[i]) {
+          const v = Number(arrLine[i]);
+          if (Number.isNaN(v)) {
+            throw new Error(`Error: invalid value line ${indexLine}.`);
```

Comment:
Again, data validation.

Reply:
#673 is only focused on validating user uploaded data, while parseFile is used for parsing server and user input files. But come to look at it, since many of the checks in this file are related to the ones created in #673, I'm starting to wonder if they should be moved in here entirely?

Reply:
Since (this PR) https://github.com/MetabolicAtlas/MetabolicAtlas/blob/0529cddd67ef8a3c5972ab0a00d7553b82421a23/frontend/src/components/explorer/mapViewer/DataOverlay.vue#L209 uses `handleErrorCustomFile` which is also used in #673 https://github.com/MetabolicAtlas/MetabolicAtlas/blob/1aeb5155e0b8442374e00e809add79aaa591c5f1/frontend/src/components/explorer/mapViewer/DataOverlay.vue#L17

Reply:
Either way, I'll think I'll leave the validation code in this file for now and work with it more when merging PR #673

Reply:
Summary from meeting (will be edited)